torch/fx/experimental/symbolic_shapes.py

```diff
@@ -88,56 +89,88 @@ def _handle_sym_dispatch(func, args, kwargs):
 
+def guard_int(a):
+    if isinstance(a, SymInt):
+        return a.node.guard_int("", 0)  # NB: uses Python backtrace
+    assert isinstance(a, int)
+    return a
+
 def sym_float(a):
-    if hasattr(a, '__sym_float__'):
-        return a.__sym_float__()
-    elif isinstance(a, torch._C.SymFloatNode):
+    if isinstance(a, SymFloat):
         return a
+    elif hasattr(a, '__sym_float__'):
+        return a.__sym_float__()
     return float(a)
 
 def sym_int(a):
-    if hasattr(a, '__sym_int__'):
-        return a.__sym_int__()
-    elif isinstance(a, torch._C.SymIntNode):
+    if isinstance(a, SymInt):
         return a
+    elif hasattr(a, '__sym_int__'):
+        return a.__sym_int__()
     return int(a)
 
 # TODO: An incomplete list
 # 1. Set variables to be equal when we do equality
 # 2. Specialize on 0/1 when we do subtraction
-class PySymInt(object):
+class SymNode:
     """
-    PySymInt objects are the primary "symbolic shape" objects that flow through
-    our program. They're what sit under FakeTensor, and contains our primary
-    implementation of symbolic shapes.
+    This is a type erased SymInt/SymFloat which we use to do actual operations.
+    End users don't touch this.  Magic methods are NOT defined on this object.
     """
-    def __init__(self, expr, shape_env, constant=None):
+    def __init__(self, expr, shape_env, pytype, constant=None):
         self._expr = expr
         self.shape_env = shape_env
+        self.pytype = pytype
         self.constant = constant
 
     @property
     def expr(self):
         self._update_expr()
         return self._expr
 
-    def wrap(self, num):
-        return PySymInt(sympy.Integer(num), self.shape_env, constant=num)
+    def _update_expr(self):
+        self._expr = self.shape_env.replace(self._expr)
+
+    def to_node(self, num):
+        if isinstance(num, (SymInt, SymFloat)):
+            return num.node
+        elif isinstance(num, int):
+            return self.wrap_int(num)
+        elif isinstance(num, float):
+            return self.wrap_float(num)
+        else:
+            # NotImplementedError is important so that Python tries the
+            # other magic method
+            raise NotImplementedError(type(num))
+
+    def is_int(self):
+        return self.pytype is int
+
+    def is_float(self):
+        return self.pytype is float
+
+    def wrap_int(self, num):
+        assert isinstance(num, int)
+        return SymNode(sympy.Integer(num), self.shape_env, int, constant=num)
+
+    def wrap_float(self, num):
+        assert isinstance(num, float)
+        return SymNode(sympy.Integer(num), self.shape_env, float, constant=num)
 
     def clone(self):
-        return PySymInt(self.expr, self.shape_env, constant=self.constant)
+        return SymNode(self.expr, self.shape_env, self.pytype, constant=self.constant)
 
-    def _update_expr(self):
-        self._expr = self.shape_env.replace(self._expr)
+    def str(self):
+        return f"{self.expr}"
```

Comment:
Do we want a `SymNode()` around? Or we will make nice printing on the SymInt/SymFloat in the future and that will be enough.

Reply:
I am ambivalent about this. On the one hand, wrapping with SymNode makes it more obvious with repr that you have a SymNode vs SymInt/SymFloat, and this is useful when debugging type confusion. OTOH, when you actually print (e.g., from C++) you really just want something clean like "2" or "s0 + s1". Let's keep it like this for now (as it preserves old behavior) and see if we want to litigate it later.